Add support for getting plaintext password from ldap for ex. digest-md5 auth.

Author: Marmarek
--- storage/authreg_ldap.c	2007-04-28 23:56:58.000000000 +0200
+++ storage/authreg_ldap.c	2007-08-05 21:55:18.000000000 +0200
@@ -113,7 +113,7 @@
       }
       
       /* starttls */
-      if(data->flags & AR_LDAP_FLAGS_STARTTLS) { 
+      if(1 || data->flags & AR_LDAP_FLAGS_STARTTLS) { 
 	if(ldap_start_tls_s(data->ld, NULL, NULL)) {
 	  log_write(data->ar->c2s->log, LOG_ERR, "ldap: couldn't start TLS: %s", ldap_err2string(_ldap_get_lderrno(data->ld)));
 	  ldap_unbind_s(data->ld);
@@ -164,9 +164,9 @@
 
 
 /** do a search, return the dn */
-static char *_ldap_search(moddata_t data, char *realm, char *username)
+static char *_ldap_search(moddata_t data, char *realm, char *username, char *password)
 {
-    char filter[1024], *dn, *no_attrs[] = { NULL }, *basedn;
+    char filter[1024], *dn, *no_attrs[] = { "userPassword", NULL }, *basedn, **vals;
     LDAPMessage *result, *entry;
 
     basedn = xhash_get(data->basedn, realm);
@@ -227,6 +227,20 @@
 
     dn = ldap_get_dn(data->ld, entry);
 
+    if (password) {
+	vals=ldap_get_values(data->ld, entry, "userPassword");
+
+	if(ldap_count_values(vals)>0) {
+	    strcpy(password, vals[0]);
+	} else {
+	    log_write(data->ar->c2s->log, LOG_ERR, "ldap: failed to get password of user %s realm %s", username, realm);
+            ldap_value_free(vals);
+	    ldap_msgfree(result);
+	    return NULL;
+	}
+	ldap_value_free(vals);
+    }
+
     ldap_msgfree(result);
 
     log_debug(ZONE, "got dn '%s' from realm '%s', user '%s'", dn, realm, username);
@@ -245,7 +259,7 @@
     do {
         xhash_iter_get((xht) ar->private, NULL, (void *) &data);
         if( ! (data->ld == NULL && _ldap_connect(data)) ) {
-            dn = _ldap_search(data, realm, username);
+            dn = _ldap_search(data, realm, username, NULL);
             if (dn != NULL) {
                 ldap_memfree(dn);
                 return 1;
@@ -273,7 +287,7 @@
 
         if( ! (data->ld == NULL && _ldap_connect(data)) ) {
 
-            dn = _ldap_search(data, realm, username);
+            dn = _ldap_search(data, realm, username, NULL);
             if (dn != NULL) {
 
                 if(ldap_simple_bind_s(data->ld, dn, password) ) {
@@ -297,6 +311,31 @@
 
 }
 
+/** get the password */
+static int _ldap_get_password(authreg_t ar, char *username, char *realm, char password[257])
+{
+
+    moddata_t data;
+    char *dn;
+
+    if(xhash_iter_first((xht) ar->private))
+    do {
+        xhash_iter_get((xht) ar->private, NULL, (void *) &data);
+
+        if( ! (data->ld == NULL && _ldap_connect(data)) ) {
+
+            dn = _ldap_search(data, realm, username, password);
+            if (dn != NULL) {
+	    	return 0;
+            }
+        }
+    } while(xhash_iter_next((xht) ar->private));
+
+
+    return 1;
+
+}
+
 /** shut me down */
 static void _ldap_free(authreg_t ar)
 {
@@ -478,6 +517,7 @@
 
     ar->user_exists = _ldap_user_exists;
     ar->check_password = _ldap_check_password;
+    ar->get_password = _ldap_get_password;
     ar->free = _ldap_free;
 
     return 0;
